In [12]:
import numpy as np
import pandas as pd

In [13]:
import plotly as py
import plotly.graph_objs as go
from plotly.offline import iplot

# Использование cufflinks в офлайн-режиме
import cufflinks
cufflinks.go_offline()

# Настройка глобальной темы cufflinks
cufflinks.set_config_file(world_readable=True, theme='pearl', offline=True)

In [14]:
filename = 'Массив чеков_ver3.csv'

Для начала прочитаем только заголовки датасета:

In [15]:
headers = pd.read_csv(filename, nrows=0, encoding='utf-16').columns.tolist()
headers

['client_NT',
 'text_kktAddress',
 'Идентификатор покупателя',
 'dateTime',
 'receipt_id',
 'CLEAN_SKU',
 'price_items',
 'quantity_items',
 'sum_items',
 'totalSum',
 'RIGHTHOLDER_NT',
 'BRAND_NT',
 'CLASS_NT',
 'GROUP_NT',
 'CATEGORY_NT',
 'SUBCATEGORY_NT',
 'BASE_NT',
 'lat',
 'lon']

Можно сформировать отдельный список всех магазинов из датасета:

In [58]:
# Сформируем список магазинов из csv-файла, прочитав только колонки 'client_NT', 'text_kktAddress', 'lat', 'lon'
stores = pd.read_csv(filename, usecols=['client_NT', 'text_kktAddress', 'lat', 'lon'], encoding='utf-16').drop_duplicates()
stores

client_NT                                    text_kktAddress  \
0               Окей                         г Москва, ул Озёрная, д 50   
7               Окей                    г Москва, пр-кт Мира, д 211 к 2   
10       Гиперглобус                    г Москва, ш Новоухтомское, д 2А   
43              Окей                 г Москва, ул Кировоградская, д 13А   
58       Гиперглобус  г Москва, Сосенское п, п Коммунарка, ул Алекса...   
109             Окей                        г Москва, ш Каширское, д 14   
117      Гиперглобус   г Москва, км Новорижское шоссе 22-й, влд 1 стр 1   
193      Гиперглобус  г Москва, км Киевское шоссе 23-й (п Московский...   
240             Окей              г Москва, ш Дмитровское, влд 163А к 1   
250             Окей                   г Москва, ш Головинское, д 5 к 1   
324      Гиперглобус                         г Москва, ул Широкая, д 2А   
338             Окей  г Москва, Московский п, г Московский, ул Хабар...   
950             Окей                     г Москва, ш Дмитровское, д 98С   
975             Окей                   г Москва, ул Святоозерская, д 1А   
3432     Гиперглобус           г Москва, км Новорижское шоссе 22-й, д 1   
5132            Окей                    г Москва, ш Дмитровское, д 163А   
26345      Metro c&c                    г Москва, пр-кт Мира, д 211 к 1   
49235    Гиперглобус            г Москва, пр-кт Волгоградский, д 42 к 9   
66968      Metro c&c                       г Москва, ул Рябиновая, д 59   
99194           Окей                     г Москва, ш Дмитровское, д 49Д   
153648     Metro c&c                     г Москва, ул Дорожная, д 1 к 1   
190584     Metro c&c                   г Москва, ш Ленинградское, д 71Г   
357187          Окей                  г Москва, ш Дмитровское, влд 163А   
591005     Metro c&c                       г Москва, км МКАД 104-й, д 6   
1172280    Metro c&c         г Москва, Московский п, кв-л 34, д 3 стр 1   
1233052         Окей                    г Москва, ш Дмитровское, д 107А   
1234331         Окей                       г Москва, ш Дмитровское, к 1   
1656407    Metro c&c              г Москва, ул Маршала Прошлякова, д 14   
1883345    Metro c&c                    г Москва, ш Дмитровское, д 165Б   
2075370         Окей                     г Москва, ш Дмитровское, д 165   
2085497         Окей                     г Москва, ш Дмитровское, д 58А   
2086659         Окей                   г Москва, ш Дмитровское, влд 91А   
2368985         Окей                      г Москва, ш Дмитровское, д 41   
2399569         Окей              г Москва, ш Дмитровское, д 107 стр 2а   
2399908         Окей               г Москва, ш Дмитровское, д 157 стр 2   

                lat         lon  
0        55,6647196  37,4372193  
7         55,845855   37,662093  
10       55,7128634  37,8454195  
43       55,6121172  37,6070486  
58        55,557266   37,494368  
109       55,664456   37,627059  
117       55,801981   37,300674  
193       55,623788   37,422027  
240        55,90968   37,540721  
250        55,84016  37,4915995  
324      55,8873966  37,6392327  
338      55,5939458  37,3518308  
950       55,879157   37,547189  
975      55,7146158  37,8835661  
3432     55,8025796  37,3019609  
5132       55,90968   37,540721  
26345     55,844041    37,66265  
49235    55,7076672   37,715361  
66968    55,6858545  37,4317811  
99194    55,8359577  37,5746979  
153648   55,6307682  37,6249431  
190584    55,875753   37,447279  
357187     55,90968   37,540721  
591005   55,8113123  37,8373703  
1172280  55,6199842  37,3874065  
1233052   55,880945    37,54118  
1234331    55,87593   37,547719  
1656407  55,7911159  37,3886894  
1883345   55,915992   37,540084  
2075370  55,9285654  37,5440174  
2085497  55,8515572  37,5663772  
2086659   55,870203   37,546893  
2368985   55,833526   37,570788  
2399569   55,878406  37,5413905  
2399908   55,898498   37,542536

In [66]:
df = pd.DataFrame(stores)
df['lat'] = df['lat'].str.replace(',', '.').astype(float)
df['lon'] = df['lon'].str.replace(',', '.').astype(float)

# Строим карту
fig = px.scatter_mapbox(
    df,
    lat='lat',
    lon='lon',
    hover_name='client_NT',
    hover_data={'text_kktAddress': True},
    zoom=10,
    height=600,
    title='Расположение магазинов'
)

# Настройки карты (используем mapbox вместо scatter_map)
fig.update_layout(
    mapbox_style="open-street-map",
    margin={"r":0,"t":40,"l":0,"b":0},
    mapbox=dict(
        center=dict(lat=df['lat'].mean(), lon=df['lon'].mean()),
        zoom=10
    )
)

# Настройки маркеров (правильный синтаксис)
fig.update_traces(
    marker=dict(
        size=14,        # Размер маркера
        color='red',    # Цвет
        opacity=0.8     # Прозрачность
    )
)

fig.show()

C:\Temp\ipykernel_32464\1226229930.py:6: DeprecationWarning:

*scatter_mapbox* is deprecated! Use *scatter_map* instead. Learn more at: https://plotly.com/python/mapbox-to-maplibre/



И отдельный список торговых сетей:

In [68]:
trades = stores['client_NT'].drop_duplicates()
trades

0               Окей
10       Гиперглобус
26345      Metro c&c
Name: client_NT, dtype: object

In [69]:
# Сохраним результат в csv файл
stores.sort_values(['client_NT', 'text_kktAddress']).to_csv('Список магазинов.csv', index=False, encoding='utf-8')
stores

client_NT                                    text_kktAddress  \
0               Окей                         г Москва, ул Озёрная, д 50   
7               Окей                    г Москва, пр-кт Мира, д 211 к 2   
10       Гиперглобус                    г Москва, ш Новоухтомское, д 2А   
43              Окей                 г Москва, ул Кировоградская, д 13А   
58       Гиперглобус  г Москва, Сосенское п, п Коммунарка, ул Алекса...   
109             Окей                        г Москва, ш Каширское, д 14   
117      Гиперглобус   г Москва, км Новорижское шоссе 22-й, влд 1 стр 1   
193      Гиперглобус  г Москва, км Киевское шоссе 23-й (п Московский...   
240             Окей              г Москва, ш Дмитровское, влд 163А к 1   
250             Окей                   г Москва, ш Головинское, д 5 к 1   
324      Гиперглобус                         г Москва, ул Широкая, д 2А   
338             Окей  г Москва, Московский п, г Московский, ул Хабар...   
950             Окей                     г Москва, ш Дмитровское, д 98С   
975             Окей                   г Москва, ул Святоозерская, д 1А   
3432     Гиперглобус           г Москва, км Новорижское шоссе 22-й, д 1   
5132            Окей                    г Москва, ш Дмитровское, д 163А   
26345      Metro c&c                    г Москва, пр-кт Мира, д 211 к 1   
49235    Гиперглобус            г Москва, пр-кт Волгоградский, д 42 к 9   
66968      Metro c&c                       г Москва, ул Рябиновая, д 59   
99194           Окей                     г Москва, ш Дмитровское, д 49Д   
153648     Metro c&c                     г Москва, ул Дорожная, д 1 к 1   
190584     Metro c&c                   г Москва, ш Ленинградское, д 71Г   
357187          Окей                  г Москва, ш Дмитровское, влд 163А   
591005     Metro c&c                       г Москва, км МКАД 104-й, д 6   
1172280    Metro c&c         г Москва, Московский п, кв-л 34, д 3 стр 1   
1233052         Окей                    г Москва, ш Дмитровское, д 107А   
1234331         Окей                       г Москва, ш Дмитровское, к 1   
1656407    Metro c&c              г Москва, ул Маршала Прошлякова, д 14   
1883345    Metro c&c                    г Москва, ш Дмитровское, д 165Б   
2075370         Окей                     г Москва, ш Дмитровское, д 165   
2085497         Окей                     г Москва, ш Дмитровское, д 58А   
2086659         Окей                   г Москва, ш Дмитровское, влд 91А   
2368985         Окей                      г Москва, ш Дмитровское, д 41   
2399569         Окей              г Москва, ш Дмитровское, д 107 стр 2а   
2399908         Окей               г Москва, ш Дмитровское, д 157 стр 2   

                lat         lon  
0        55,6647196  37,4372193  
7         55,845855   37,662093  
10       55,7128634  37,8454195  
43       55,6121172  37,6070486  
58        55,557266   37,494368  
109       55,664456   37,627059  
117       55,801981   37,300674  
193       55,623788   37,422027  
240        55,90968   37,540721  
250        55,84016  37,4915995  
324      55,8873966  37,6392327  
338      55,5939458  37,3518308  
950       55,879157   37,547189  
975      55,7146158  37,8835661  
3432     55,8025796  37,3019609  
5132       55,90968   37,540721  
26345     55,844041    37,66265  
49235    55,7076672   37,715361  
66968    55,6858545  37,4317811  
99194    55,8359577  37,5746979  
153648   55,6307682  37,6249431  
190584    55,875753   37,447279  
357187     55,90968   37,540721  
591005   55,8113123  37,8373703  
1172280  55,6199842  37,3874065  
1233052   55,880945    37,54118  
1234331    55,87593   37,547719  
1656407  55,7911159  37,3886894  
1883345   55,915992   37,540084  
2075370  55,9285654  37,5440174  
2085497  55,8515572  37,5663772  
2086659   55,870203   37,546893  
2368985   55,833526   37,570788  
2399569   55,878406  37,5413905  
2399908   55,898498   37,542536

Переопределим некоторые типы данных для экономии места:

In [19]:
# Здесь сделано под ver2, но под ver3 тоже подходит, так как в ver3 меньше столбцов, а остальные названы также
dtype_spec = {
    'region': 'category',
    'userInn': 'int64',
    'client_NT': 'category',
    'text_kktAddress': 'category',
    'Идентификатор покупателя': 'str',
    'dateTime': 'str', # после преобразуем в datetime
    'Year': 'int16',
    'Month': 'int16',
    'receipt_id': 'str',
    'CLEAN_SKU': 'str',
    'price_items': 'float32',
    'quantity_items': 'float32',
    'sum_items': 'float32',
    'totalSum': 'float32',
    'RIGHTHOLDER_NT': 'str',
    'BRAND_NT': 'category',
    'CLASS_NT': 'category',
    'GROUP_NT': 'category',
    'CATEGORY_NT': 'category',
    'SUBCATEGORY_NT': 'category',
    'BASE_NT': 'category',
    'lat': 'float32',
    'lon': 'float32'
}

Исходя из примера массива, повторяющиеся строки являются частью одного чека. Значит дубликаты строк не убираем

Загружаем датасет:

In [20]:
data = pd.read_csv(filename, encoding='utf-16', dtype=dtype_spec, decimal=',').reset_index()

Если появляется ошибка памяти, то можно загрузить датасет частями:

In [21]:
# data = pd.DataFrame()
# chunk_size = 1000000
# for chunk in pd.read_csv('Массив чеков_ver2.csv', encoding='utf-16', dtype=dtype_spec, decimal=',', chunksize=chunk_size):
#     data = pd.concat([data, chunk], ignore_index=True)
# data  = data.reset_index()

Посмотрим первые 5 строк, чтобы убедиться, что всё подгрузилось правильно:

In [22]:
data.head(5)

index client_NT             text_kktAddress Идентификатор покупателя  \
0      0      Окей  г Москва, ул Озёрная, д 50                      NaN   
1      1      Окей  г Москва, ул Озёрная, д 50                      NaN   
2      2      Окей  г Москва, ул Озёрная, д 50                      NaN   
3      3      Окей  г Москва, ул Озёрная, д 50                      NaN   
4      4      Окей  г Москва, ул Озёрная, д 50                      NaN   

              dateTime              receipt_id          CLEAN_SKU  \
0  26.06.2022 13:51:00  9960440300975081:17128          Бананы кг   
1  26.06.2022 13:46:00  9960440300975081:17126  Зажарка Maggi 60г   
2  01.05.2022 13:26:00  9960440300975081:11611          Бананы кг   
3  01.05.2022 13:26:00  9960440300975081:11611          Яблоки кг   
4  04.06.2022 18:19:00   9960440301010969:5365       Апельсины кг   

   price_items  quantity_items  sum_items  totalSum RIGHTHOLDER_NT BRAND_NT  \
0        70.38            0.59      41.81   2135.82            NaN      NaN   
1        33.42            1.00      33.42   3920.33         NESTLE    MAGGI   
2        96.24            0.91      87.58   2387.77            NaN      NaN   
3        87.49            1.30     113.74   2387.77            NaN      NaN   
4       109.99            1.69     186.10  13641.60            NaN      NaN   

  CLASS_NT         GROUP_NT                        CATEGORY_NT  \
0     ФРОВ    СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫЕ   
1     БАКЛ  БАКАЛЕЯ СЫПУЧАЯ  ПРИПРАВЫ ДЛЯ ПЕРВЫХ И ВТОРЫХ БЛЮД   
2     ФРОВ    СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫЕ   
3     ФРОВ    СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫЕ   
4      NaN              NaN                                NaN   

        SUBCATEGORY_NT              BASE_NT   lat   lon  
0               БАНАНЫ                БАНАН 55.66 37.44  
1  ПРИПРАВА ДЛЯ ЖАРКОЕ  ПРИПРАВА ДЛЯ ЖАРКОЕ 55.66 37.44  
2               БАНАНЫ                БАНАН 55.66 37.44  
3               ЯБЛОКИ               ЯБЛОКО 55.66 37.44  
4                  NaN                  NaN 55.66 37.44

Проверим данные, например, чек номер `9960440300975081:11611`:

In [23]:
check = data[data['receipt_id'] == '9960440300975081:11611']
check

index client_NT             text_kktAddress  \
2              2      Окей  г Москва, ул Озёрная, д 50   
3              3      Окей  г Москва, ул Озёрная, д 50   
1342543  1342543      Окей  г Москва, ул Озёрная, д 50   
1342544  1342544      Окей  г Москва, ул Озёрная, д 50   
1342545  1342545      Окей  г Москва, ул Озёрная, д 50   
1342546  1342546      Окей  г Москва, ул Озёрная, д 50   
1342547  1342547      Окей  г Москва, ул Озёрная, д 50   
1342548  1342548      Окей  г Москва, ул Озёрная, д 50   
1342549  1342549      Окей  г Москва, ул Озёрная, д 50   
1342550  1342550      Окей  г Москва, ул Озёрная, д 50   
1342551  1342551      Окей  г Москва, ул Озёрная, д 50   
1342552  1342552      Окей  г Москва, ул Озёрная, д 50   
1342553  1342553      Окей  г Москва, ул Озёрная, д 50   
1342554  1342554      Окей  г Москва, ул Озёрная, д 50   
1342555  1342555      Окей  г Москва, ул Озёрная, д 50   
1342556  1342556      Окей  г Москва, ул Озёрная, д 50   
1342557  1342557      Окей  г Москва, ул Озёрная, д 50   
1342558  1342558      Окей  г Москва, ул Озёрная, д 50   
1342559  1342559      Окей  г Москва, ул Озёрная, д 50   
1342560  1342560      Окей  г Москва, ул Озёрная, д 50   
1342561  1342561      Окей  г Москва, ул Озёрная, д 50   
1342562  1342562      Окей  г Москва, ул Озёрная, д 50   
1342563  1342563      Окей  г Москва, ул Озёрная, д 50   
1342564  1342564      Окей  г Москва, ул Озёрная, д 50   
1342565  1342565      Окей  г Москва, ул Озёрная, д 50   
1342566  1342566      Окей  г Москва, ул Озёрная, д 50   
1342567  1342567      Окей  г Москва, ул Озёрная, д 50   

        Идентификатор покупателя             dateTime              receipt_id  \
2                            NaN  01.05.2022 13:26:00  9960440300975081:11611   
3                            NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342543                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342544                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342545                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342546                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342547                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342548                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342549                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342550                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342551                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342552                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342553                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342554                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342555                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342556                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342557                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342558                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342559                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342560                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342561                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342562                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342563                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342564                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342565                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342566                      NaN  01.05.2022 13:26:00  9960440300975081:11611   
1342567                      NaN  01.05.2022 13:26:00  9960440300975081:11611   

                                                 CLEAN_SKU  price_items  \
2                                         

Проверим, что сумма по столбцу `sum_items` равна значению `totalSum`:

In [24]:
check['sum_items'].sum()

np.float32(2387.77)

Посмотрим общую информацию о датафрейме:

In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2400099 entries, 0 to 2400098
Data columns (total 20 columns):
 #   Column                    Dtype   
---  ------                    -----   
 0   index                     int64   
 1   client_NT                 category
 2   text_kktAddress           category
 3   Идентификатор покупателя  object  
 4   dateTime                  object  
 5   receipt_id                object  
 6   CLEAN_SKU                 object  
 7   price_items               float32 
 8   quantity_items            float32 
 9   sum_items                 float32 
 10  totalSum                  float32 
 11  RIGHTHOLDER_NT            object  
 12  BRAND_NT                  category
 13  CLASS_NT                  category
 14  GROUP_NT                  category
 15  CATEGORY_NT               category
 16  SUBCATEGORY_NT            category
 17  BASE_NT                   category
 18  lat                       float32 
 19  lon                       float32 
dtypes:

Посмотрим описательные статистики для числовых столбцов:

In [26]:
data.describe()

index  price_items  quantity_items  sum_items   totalSum  \
count 2400099.00   2400099.00      2400099.00 2400099.00 2400099.00   
mean  1200049.00       175.96            1.33     178.95    4624.64   
std    692849.05       238.92            1.82     260.98    3830.94   
min         0.00         0.00            0.00       0.00       0.00   
25%    600024.50        59.39            1.00      65.69    2309.50   
50%   1200049.00       102.51            1.00     115.00    3680.91   
75%   1800073.50       199.00            1.00     209.98    5748.90   
max   2400098.00     38986.09          400.00   40499.00  122134.65   

             lat        lon  
count 2400099.00 2400099.00  
mean       55.73      37.52  
std         0.12       0.17  
min        55.56      37.30  
25%        55.62      37.42  
50%        55.71      37.49  
75%        55.84      37.64  
max        55.93      37.88

Посмотрим корреляционные отношения между числовыми столбцами

In [27]:
corr_matrix = data[['price_items', 'quantity_items', 'sum_items', 'totalSum', 'lat', 'lon']].corr()

fig = go.Figure(data=go.Heatmap(
    z=corr_matrix.values,
    x=corr_matrix.columns,
    y=corr_matrix.index,
    zmin=-1, 
    zmax=1, 
    colorscale='RdBu', 
    text=np.round(corr_matrix.values, 2), 
    texttemplate="%{text}",
    hoverinfo="x+y+z",
    showscale=True 
))

# Настройки макета
fig.update_layout(
    title='Корреляционная матрица (интерактивная)',
    xaxis_title='Переменные',
    yaxis_title='Переменные',
    xaxis_showgrid=False,
    yaxis_showgrid=False,
    yaxis_autorange='reversed',  # реверсируем ось Y как в sns
    width=800,
    height=600
)

# Дополнительные настройки аннотаций
fig.update_traces(
    textfont_size=12,
    textfont_color='black'
)

fig.show()

Практически везде слабая корреляционная зависимость, за исключением зависимости sum_items - price_items, она средняя(0.62).

Добавим новые столбцы hour и weekday. И переименуем столбец "Идентификатор покупателя" в customer_id

In [28]:
data['hour'] = pd.to_datetime(data.dateTime).dt.hour

weekdays = {
    0: '1.Понедельник',
    1: '2.Вторник',
    2: '3.Среда',
    3: '4.Четверг',
    4: '5.Пятница',
    5: '6.Суббота',
    6: '7.Воскресенье'
    }
data['weekday'] = pd.to_datetime(data.dateTime).dt.weekday.map(weekdays)

data = data.rename(columns={'Идентификатор покупателя': 'customer_id'})

C:\Temp\ipykernel_32464\2744415287.py:1: UserWarning:

Parsing dates in %d.%m.%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.

C:\Temp\ipykernel_32464\2744415287.py:12: UserWarning:

Parsing dates in %d.%m.%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.



Проведём `ABC-анализ` для каждой торговой сети.

Классифицируем товары `CLEAN_SKU` по прибыли и количеству проданных позиций. А так же по `CATEGORY_NT` по количеству проданных позиций

In [29]:
def perform_abc(df, index):
    cols = list(df.columns)
    for i in index:
        cols.remove(i)
    groupped_df = df.groupby(index[1]).agg({col: 'sum' for col in cols})
    for col in cols:
        groupped_df[f'rel_{col}'] = groupped_df[col] / sum(groupped_df[col])
        groupped_df = groupped_df.sort_values(f'rel_{col}', ascending=False)
        groupped_df[f'cumsum_{col}'] = groupped_df[f'rel_{col}'].cumsum()
        groupped_df[f'abc_{col}'] = np.where(groupped_df[f'cumsum_{col}'] < 0.8, 'A', np.where(groupped_df[f'cumsum_{col}'] < 0.95, 'B', 'C'))
    return groupped_df[cols + [f'abc_{col}' for col in cols]]

In [30]:
abc_all_SKU = data[['client_NT', 'CLEAN_SKU', 'price_items', 'quantity_items']].dropna()
abc_all_SKU

client_NT                                          CLEAN_SKU  \
0            Окей                                          Бананы кг   
1            Окей                                  Зажарка Maggi 60г   
2            Окей                                          Бананы кг   
3            Окей                                          Яблоки кг   
4            Окей                                       Апельсины кг   
...           ...                                                ...   
2400094      Окей  Кондиционер конц д/белья Vernel supreme elegan...   
2400095      Окей                                 Абрикосы 500г упак   
2400096      Окей   Пакет подарочный Арт и Дизайн ST премиум 12х36см   
2400097      Окей  Вино Вадиаперти Греко Ди Туфо бел. сух. 4,5-16...   
2400098      Окей                    Набор конфет Merci ассорти 400г   

         price_items  quantity_items  
0              70.38            0.59  
1              33.42            1.00  
2              96.24            0.91  
3              87.49            1.30  
4             109.99            1.69  
...              ...             ...  
2400094       199.99            1.00  
2400095       169.99            1.00  
2400096       102.99            1.00  
2400097      1099.02            1.00  
2400098       629.99            1.00  

[2400064 rows x 4 columns]

In [31]:
abc_SKU = pd.DataFrame()
for trade in trades:
    tmp = pd.DataFrame()
    tmp = perform_abc(abc_all_SKU[abc_all_SKU['client_NT'] == trade], ['client_NT', 'CLEAN_SKU']).reset_index()
    tmp['client_NT'] = trade
    abc_SKU = pd.concat([abc_SKU, tmp], ignore_index=True).dropna()
abc_SKU

CLEAN_SKU  price_items  \
0      Пакет из переработанного материала Kinglet 40х...     67268.21   
1                                              Бананы кг    652318.31   
2         БЗМЖ Молоко утп Домик в деревне 3,2% 925мл тба    312110.16   
3                          Картофель ранний импортный кг    252049.86   
4      Сахар-песок бел. кристал. ГОСТ 33222-2015, кат...    199892.50   
...                                                  ...          ...   
71228                 630Г КОЛБАСА ОКРАИНА ЗЕРНИСТАЯ С/К       565.59   
71229                                      САЛАТ АЙСБЕРГ       159.01   
71230                                             ЧЕСНОК       196.96   
71231                                              ВИШНЯ       239.00   
71232             350Г КАРПАЧЧО ИЗ МЯСА ПТИЦЫ РУБЛЕВСКИЙ       827.44   

       quantity_items abc_price_items abc_quantity_items  client_NT  
0             9074.00               A                  A       Окей  
1             8526.12               A                  A       Окей  
2             7176.00               A                  A       Окей  
3             6234.26               A                  A       Окей  
4             5130.00               A                  A       Окей  
...               ...             ...                ...        ...  
71228            0.34               A                  C  Metro c&c  
71229            0.31               B                  C  Metro c&c  
71230            0.31               B                  C  Metro c&c  
71231            0.21               B                  C  Metro c&c  
71232            0.12               A                  C  Metro c&c  

[71233 rows x 6 columns]

In [32]:
abc_all_CATEGORY = data[['client_NT', 'CATEGORY_NT', 'quantity_items']].dropna()
abc_all_CATEGORY

client_NT                        CATEGORY_NT  quantity_items
0            Окей                    ФРУКТЫ ОСНОВНЫЕ            0.59
1            Окей  ПРИПРАВЫ ДЛЯ ПЕРВЫХ И ВТОРЫХ БЛЮД            1.00
2            Окей                    ФРУКТЫ ОСНОВНЫЕ            0.91
3            Окей                    ФРУКТЫ ОСНОВНЫЕ            1.30
5            Окей                       ОВОЩИ РАЗНЫЕ            0.41
...           ...                                ...             ...
2400094      Окей                   СР-ВА ДЛЯ СТИРКИ            1.00
2400095      Окей                    ФРУКТЫ ОСНОВНЫЕ            1.00
2400096      Окей         ВСЕ ДЛЯ ВЕСЕЛОГО ПРАЗДНИКА            1.00
2400097      Окей                               ВИНО            1.00
2400098      Окей                НАБОРЫ КОНДИТЕРСКИЕ            1.00

[1626605 rows x 3 columns]

In [33]:
abc_CATEGORY = pd.DataFrame()
for trade in trades:
    tmp = pd.DataFrame()
    tmp = perform_abc(abc_all_CATEGORY[abc_all_CATEGORY['client_NT'] == trade], ['client_NT', 'CATEGORY_NT']).reset_index()
    tmp['client_NT'] = trade
    abc_CATEGORY = pd.concat([abc_CATEGORY, tmp], ignore_index=True).dropna()
abc_CATEGORY = abc_CATEGORY.rename(columns={'abc_quantity_items': 'abc_quantity_items_CAT'})
abc_CATEGORY


C:\Temp\ipykernel_32464\3440270023.py:5: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

C:\Temp\ipykernel_32464\3440270023.py:5: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

C:\Temp\ipykernel_32464\3440270023.py:5: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



CATEGORY_NT  quantity_items  \
0                       ЗООКОРМ, ЛАКОМСТВА ДЛЯ КОШЕК        51633.00   
1                                   ТВОРОЖНАЯ ГРУППА        29286.00   
2                                               ВОДА        29043.00   
3                                             МОЛОКО        28928.00   
4                                         КОРНЕПЛОДЫ        28415.69   
...                                              ...             ...   
1279                         ШАРФЫ, ПАЛАНТИНЫ, ПАРЕО            0.00   
1280                                            САКЕ            0.00   
1281                                         БАЛЬЗАМ            0.00   
1282                   НОСИТЕЛИ ИНФОРМАЦИИ С ЗАПИСЬЮ            0.00   
1283  АКСЕССУАРЫ ДЛЯ ПРОГУЛКИ И ДРЕССИРОВКИ ЖИВОТНЫХ            0.00   

     abc_quantity_items_CAT  client_NT  
0                         A       Окей  
1                         A       Окей  
2                         A       Окей  
3                         A       Окей  
4                         A       Окей  
...                     ...        ...  
1279                      C  Metro c&c  
1280                      C  Metro c&c  
1281                      C  Metro c&c  
1282                      C  Metro c&c  
1283                      C  Metro c&c  

[1284 rows x 4 columns]

In [34]:
xyz_data = data[['client_NT', 'CLEAN_SKU', 'dateTime', 'quantity_items']]
xyz_data['dateTime'] = pd.to_datetime(xyz_data['dateTime']).dt.strftime('%W') # По неделям
xyz_data

C:\Temp\ipykernel_32464\347484163.py:2: UserWarning:

Parsing dates in %d.%m.%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.



client_NT                                          CLEAN_SKU dateTime  \
0            Окей                                          Бананы кг       25   
1            Окей                                  Зажарка Maggi 60г       25   
2            Окей                                          Бананы кг       17   
3            Окей                                          Яблоки кг       17   
4            Окей                                       Апельсины кг       22   
...           ...                                                ...      ...   
2400094      Окей  Кондиционер конц д/белья Vernel supreme elegan...       20   
2400095      Окей                                 Абрикосы 500г упак       20   
2400096      Окей   Пакет подарочный Арт и Дизайн ST премиум 12х36см       21   
2400097      Окей  Вино Вадиаперти Греко Ди Туфо бел. сух. 4,5-16...       21   
2400098      Окей                    Набор конфет Merci ассорти 400г       21   

         quantity_items  
0                  0.59  
1                  1.00  
2                  0.91  
3                  1.30  
4                  1.69  
...                 ...  
2400094            1.00  
2400095            1.00  
2400096            1.00  
2400097            1.00  
2400098            1.00  

[2400099 rows x 4 columns]

In [35]:
def perform_xyz(df, date_col, index_cols, value_col):
    # Группировка и агрегация (сумма quantity_items по неделям)
    grouped = df.groupby(index_cols + [date_col])[value_col].sum().reset_index()
    # Фильтрация SKU с достаточным количеством недель (>1)
    sku_week_counts = grouped.groupby(index_cols)[date_col].count().reset_index()
    valid_skus = sku_week_counts[sku_week_counts[date_col] > 1][index_cols]
    filtered = pd.merge(grouped, valid_skus, on=index_cols)
    
    # Расчет коэффициента вариации
    cv_result = (filtered.groupby(index_cols)[value_col]
                 .apply(lambda x: x.std()/x.mean())
                 .reset_index(name='cv'))
    
    # 6. Классификация XYZ
    cv_result['XYZ'] = np.where(cv_result['cv'] < 0.1, 'X', np.where(cv_result['cv'] < 0.25, 'Y', 'Z'))
    
    return cv_result

In [36]:
xyz_SKU = pd.DataFrame()
for trade in trades:
    tmp = pd.DataFrame()
    tmp = perform_xyz(xyz_data[xyz_data['client_NT'] == trade], date_col='dateTime', index_cols=['CLEAN_SKU'], value_col='quantity_items')
    tmp['client_NT'] = trade
    xyz_SKU = pd.concat([xyz_SKU, tmp], ignore_index=True).dropna()
xyz_SKU

CLEAN_SKU   cv XYZ  client_NT
0                   "Ассорти ""Черри и корнишоны"" 500г" 0.00   X       Окей
1                    "Ассорти овощное ""Пятерка"" 1000г" 0.47   Z       Окей
2      "БЗМЖ сыр ""Моне"" с массовой долей жира 45% (... 0.34   Z       Окей
3      "Бальзам Травяной 250 мл ""Живая природа"" (Ок... 0.29   Z       Окей
4      "Батончик мюсли ""О'кей"" вишня с цельными зла... 0.77   Z       Окей
...                                                  ...  ...  ..        ...
52672                                    ТОМАТЫ ВЕС (КГ) 0.81   Z  Metro c&c
52673                     ТУШКА ЦБ СТМ MC ПАКЕТ ОХЛ (КГ) 0.01   X  Metro c&c
52674         ФИЛЕ КУРИНОЕ METRO CHEF МИНИКОРОБ ОХЛ (КГ) 0.05   X  Metro c&c
52675                        ЯБЛОКИ ГРЕННИ СМИТ ВЕС (КГ) 0.52   Z  Metro c&c
52676                  ЯЙЦО КУР METRO CHEF 30 ШТ С1 (ШТ) 0.00   X  Metro c&c

[52677 rows x 4 columns]

Теперь соединим результаты abc_SKU, abc_CATEGORY, xyz_SKU в один датафрейм и выгрузим в pickle-файл

In [37]:
analysis_df = pd.DataFrame()
analysis_df = pd.merge(data, abc_SKU, on=['client_NT', 'CLEAN_SKU'], how='left')
analysis_df = pd.merge(analysis_df, abc_CATEGORY, on=['client_NT', 'CATEGORY_NT'], how='left')
analysis_df = pd.merge(analysis_df, xyz_SKU, on=['client_NT', 'CLEAN_SKU'], how='left')
analysis_df = analysis_df[['client_NT', 'text_kktAddress', 
                            'customer_id', 'dateTime', 'receipt_id', 'CLEAN_SKU', 'price_items_x',
                            'abc_price_items', 'abc_quantity_items', 'abc_quantity_items_CAT', 'XYZ', 
                            'quantity_items', 'sum_items', 'totalSum',
                            'RIGHTHOLDER_NT', 'BRAND_NT', 'CLASS_NT', 'GROUP_NT', 'CATEGORY_NT',
                            'SUBCATEGORY_NT', 'BASE_NT', 'lat', 'lon']]
analysis_df

client_NT                     text_kktAddress  \
0            Окей          г Москва, ул Озёрная, д 50   
1            Окей          г Москва, ул Озёрная, д 50   
2            Окей          г Москва, ул Озёрная, д 50   
3            Окей          г Москва, ул Озёрная, д 50   
4            Окей          г Москва, ул Озёрная, д 50   
...           ...                                 ...   
2400094      Окей  г Москва, ул Кировоградская, д 13А   
2400095      Окей  г Москва, ул Кировоградская, д 13А   
2400096      Окей  г Москва, ул Кировоградская, д 13А   
2400097      Окей  г Москва, ул Кировоградская, д 13А   
2400098      Окей  г Москва, ул Кировоградская, д 13А   

                              customer_id             dateTime  \
0                                     NaN  26.06.2022 13:51:00   
1                                     NaN  26.06.2022 13:46:00   
2                                     NaN  01.05.2022 13:26:00   
3                                     NaN  01.05.2022 13:26:00   
4                                     NaN  04.06.2022 18:19:00   
...                                   ...                  ...   
2400094                               NaN  21.05.2022 18:02:00   
2400095                               NaN  21.05.2022 18:02:00   
2400096  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   
2400097  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   
2400098  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   

                     receipt_id  \
0        9960440300975081:17128   
1        9960440300975081:17126   
2        9960440300975081:11611   
3        9960440300975081:11611   
4         9960440301010969:5365   
...                         ...   
2400094  9960440300963884:53239   
2400095  9960440300963884:53239   
2400096  9960440300964629:76254   
2400097  9960440300964629:76254   
2400098  9960440300964629:76254   

                                                 CLEAN_SKU  price_items_x  \
0                                                Бананы кг          70.38   
1                                        Зажарка Maggi 60г          33.42   
2                                                Бананы кг          96.24   
3                                                Яблоки кг          87.49   
4                                             Апельсины кг         109.99   
...                                                    ...            ...   
2400094  Кондиционер конц д/белья Vernel supreme elegan...         199.99   
2400095                                 Абрикосы 500г упак         169.99   
2400096   Пакет подарочный Арт и Дизайн ST премиум 12х36см         102.99   
2400097  Вино Вадиаперти Греко Ди Туфо бел. сух. 4,5-16...        1099.02   
2400098                    Набор конфет Merci ассорти 400г         629.99   

        abc_price_items abc_quantity_items abc_quantity_items_CAT  ...  \
0                     A                  A                      A  ...   
1                     B                  A                      A  ...   
2                     A                  A                      A  ...   
3                     A                  A                      A  ...   
4                     A                  A                    NaN  ...   
...                 ...                ...                    ...  ...   
2400094               B                  B                      A  ...   
2400095               A                  A                      A  ...   
2400096               C                  C                      C  ...   
2400097               B                  C                      B  ...   
2400098               A                  B                      B  ...   

        totalSum   RIGHTHOLDER_NT      BRAND_NT  CLASS_NT         GROUP_NT  \
0        2135.82              NaN           NaN      ФРОВ    СВЕЖИЕ ФРУКТЫ   
1        3920.33           NESTLE         MAGGI      БАКЛ  БАКАЛЕЯ СЫПУЧАЯ   
2        2387.77              NaN           NaN      ФРОВ    

In [52]:
# Сохраним результаты ABC|XYZ-анализа в pcl файл
analysis_df.to_pickle('ABC-XYZ.pcl.gz', compression='gzip')
# Можно скачать его https://disk.yandex.ru/d/aVAb0DY98upb3w

In [53]:
# Если надо загрузить, то можно воспользоваться:
# abc_xyz = pd.read_pickle('ABC-XYZ.pcl.gz', compression='gzip')

Сформируем csv-файлы по торговым сетям

In [54]:
for trade in trades:
    analysis_df[analysis_df['client_NT'] == trade].to_csv(f'ABC-XYZ {trade}.csv', index=False)

Сравним цены на одинаковые товары в разных торговых сетях.

Составим сводную таблицу по категориям товара (`CATEGORY_NT`). Столбцами будут торговые сети, а в качестве значений будет средняя цена по категории.

In [55]:
data.pivot_table(
    index='CATEGORY_NT',
    columns='client_NT',
    values='price_items',
    aggfunc='mean', 
    observed=False
).sort_index()

client_NT                                           Metro c&c  Гиперглобус  \
CATEGORY_NT                                                                  
АКСЕССУАРИКА                                              NaN       389.48   
АКСЕССУАРЫ                                                NaN       215.04   
АКСЕССУАРЫ ДЛЯ ОБУВИ                                      NaN       175.51   
АКСЕССУАРЫ ДЛЯ ПРОГУЛКИ И ДРЕССИРОВКИ ЖИВОТНЫХ            NaN       211.44   
АССОРТИ МЯСНОЙ ГАСТРОНОМИИ                                NaN       377.21   
...                                                       ...          ...   
СИГАРЫ, СИГАРИЛЫ                                          NaN       200.00   
НАКЛАДКИ, ВКЛАДЫШИ, ПРОКЛАДКИ ДЛЯ БЕРЕМЕННЫХ И ...        NaN       244.00   
МЯСО УТКИ ЗАМОРОЖЕННОЕ                                    NaN          NaN   
РУБАШКИ, БЛУЗКИ ДЕТСКИЕ                                   NaN          NaN   
ПУТЕВОДИТЕЛИ                                              NaN          NaN   

client_NT                                            Окей  
CATEGORY_NT                                                
АКСЕССУАРИКА                                       429.44  
АКСЕССУАРЫ                                         228.87  
АКСЕССУАРЫ ДЛЯ ОБУВИ                               210.36  
АКСЕССУАРЫ ДЛЯ ПРОГУЛКИ И ДРЕССИРОВКИ ЖИВОТНЫХ     258.78  
АССОРТИ МЯСНОЙ ГАСТРОНОМИИ                         151.43  
...                                                   ...  
СИГАРЫ, СИГАРИЛЫ                                      NaN  
НАКЛАДКИ, ВКЛАДЫШИ, ПРОКЛАДКИ ДЛЯ БЕРЕМЕННЫХ И ...    NaN  
МЯСО УТКИ ЗАМОРОЖЕННОЕ                             346.86  
РУБАШКИ, БЛУЗКИ ДЕТСКИЕ                            193.02  
ПУТЕВОДИТЕЛИ                                       738.32  

[428 rows x 3 columns]

Так же в этой таблице видны категории товаров, которые присутствуют в одной сети, но отсутствуют в другой

Можно посмотреть различие цен на товары, которые присутствуют во всех торговых сетях

In [56]:
data.pivot_table(
    index='CATEGORY_NT',
    columns='client_NT',
    values='price_items',
    aggfunc='mean', 
    observed=False
).sort_index().dropna()

client_NT              Metro c&c  Гиперглобус   Окей
CATEGORY_NT                                         
БАГЕТ                     105.85        50.79  27.36
БАРАНОЧНЫЕ ИЗДЕЛИЯ         43.78        55.19  48.15
БАТАРЕЙКИ                 259.00       212.60 204.88
БЕЗРЕЦЕПТУРНАЯ АПТЕКА     356.06       219.57 424.06
БЛЮДА ДЛЯ ПЕРЕКУСА        195.80       198.76 181.43
...                          ...          ...    ...
ШОКОЛАДНЫЕ ЯЙЦА            88.26       108.87  93.64
ЯГОДЫ                     204.84       302.72 194.01
ЯЙЦА ПРОЧИЕ               349.07       163.12 111.38
ТОВАРЫ ДЛЯ ЗАГОТОВОК      153.00       277.44 161.19
БРЕНДИ                    999.00       587.19 738.97

[199 rows x 3 columns]

In [57]:
data

index client_NT                     text_kktAddress  \
0              0      Окей          г Москва, ул Озёрная, д 50   
1              1      Окей          г Москва, ул Озёрная, д 50   
2              2      Окей          г Москва, ул Озёрная, д 50   
3              3      Окей          г Москва, ул Озёрная, д 50   
4              4      Окей          г Москва, ул Озёрная, д 50   
...          ...       ...                                 ...   
2400094  2400094      Окей  г Москва, ул Кировоградская, д 13А   
2400095  2400095      Окей  г Москва, ул Кировоградская, д 13А   
2400096  2400096      Окей  г Москва, ул Кировоградская, д 13А   
2400097  2400097      Окей  г Москва, ул Кировоградская, д 13А   
2400098  2400098      Окей  г Москва, ул Кировоградская, д 13А   

                              customer_id             dateTime  \
0                                     NaN  26.06.2022 13:51:00   
1                                     NaN  26.06.2022 13:46:00   
2                                     NaN  01.05.2022 13:26:00   
3                                     NaN  01.05.2022 13:26:00   
4                                     NaN  04.06.2022 18:19:00   
...                                   ...                  ...   
2400094                               NaN  21.05.2022 18:02:00   
2400095                               NaN  21.05.2022 18:02:00   
2400096  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   
2400097  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   
2400098  8e72f68f3891ce5f91503cb8dbe1b210  26.05.2022 18:08:00   

                     receipt_id  \
0        9960440300975081:17128   
1        9960440300975081:17126   
2        9960440300975081:11611   
3        9960440300975081:11611   
4         9960440301010969:5365   
...                         ...   
2400094  9960440300963884:53239   
2400095  9960440300963884:53239   
2400096  9960440300964629:76254   
2400097  9960440300964629:76254   
2400098  9960440300964629:76254   

                                                 CLEAN_SKU  price_items  \
0                                                Бананы кг        70.38   
1                                        Зажарка Maggi 60г        33.42   
2                                                Бананы кг        96.24   
3                                                Яблоки кг        87.49   
4                                             Апельсины кг       109.99   
...                                                    ...          ...   
2400094  Кондиционер конц д/белья Vernel supreme elegan...       199.99   
2400095                                 Абрикосы 500г упак       169.99   
2400096   Пакет подарочный Арт и Дизайн ST премиум 12х36см       102.99   
2400097  Вино Вадиаперти Греко Ди Туфо бел. сух. 4,5-16...      1099.02   
2400098                    Набор конфет Merci ассорти 400г       629.99   

         quantity_items  sum_items  ...      BRAND_NT CLASS_NT  \
0                  0.59      41.81  ...           NaN     ФРОВ   
1                  1.00      33.42  ...         MAGGI     БАКЛ   
2                  0.91      87.58  ...           NaN     ФРОВ   
3                  1.30     113.74  ...           NaN     ФРОВ   
4                  1.69     186.10  ...           NaN      NaN   
...                 ...        ...  ...           ...      ...   
2400094            1.00     199.99  ...        VERNEL     КЗБХ   
2400095            1.00     169.99  ...           NaN     ФРОВ   
2400096            1.00     102.99  ...  АРТ И ДИЗАЙН     КЦИГ   
2400097            1.00    1099.02  ...           NaN     АЛКГ   
2400098            1.00     629.99  ...         MERCI     КНХЛ   

                GROUP_NT                        CATEGORY_NT  \
0          СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫЕ   
1        БАКАЛЕЯ СЫПУЧАЯ  ПРИПРАВЫ ДЛЯ ПЕРВЫХ И ВТОРЫХ БЛЮД   
2          СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫЕ   
3          СВЕЖИЕ ФРУКТЫ                    ФРУКТЫ ОСНОВНЫ

Подготовим таблицу для среднего чека

In [38]:
mean_check = data.groupby('receipt_id').agg({
    'totalSum': 'max',
    'client_NT': 'first',
    'text_kktAddress': 'first',
    'hour': 'first',
    'weekday': 'first'
    
}).reset_index()

Средний чек в каждой сети

In [40]:
mc1 = mean_check.pivot_table(
    columns='client_NT',
    values='totalSum',
    aggfunc='mean', 
    observed=False
)
mc1

client_NT  Metro c&c  Гиперглобус    Окей
totalSum     8730.40      3360.11 2496.69

Сформируем таблицу со средним чеком по часам во всех торговых сетях для сравнения:

In [41]:
mc2 =mean_check.pivot_table(
    values='totalSum',
    index='hour', 
    columns='client_NT', 
    aggfunc='mean', 
    observed=False
    )
mc2

client_NT  Metro c&c  Гиперглобус    Окей
hour                                     
0                NaN      2376.18 2569.96
1                NaN      2623.16     NaN
2                NaN      2407.63     NaN
3                NaN      2262.54     NaN
4                NaN      2684.25     NaN
5            7796.83      2627.15     NaN
6            1856.69      2821.99     NaN
7           23387.27      3177.00     NaN
8           16492.82      3633.36 2417.80
9           12326.93      3589.17 2643.82
10          11505.26      3504.46 2715.09
11           4664.98      3433.57 2580.23
12           8866.72      3469.34 2520.44
13          10651.90      3404.62 2600.92
14          10963.77      3462.99 2675.17
15           7569.85      3381.50 2538.04
16           5351.76      3322.11 2487.18
17           5216.42      3329.20 2452.79
18           9440.87      3215.74 2493.00
19           7129.12      3161.01 2459.70
20           5645.90      3125.37 2392.78
21          10829.43      3146.57 2075.17
22           5457.23      3030.57 1903.43
23           3892.53      2821.54 2190.20

In [56]:
df_long = mc2.reset_index().melt(id_vars='hour', var_name='client_NT', value_name='mean_totalSum')

# Строим barplot
fig = px.bar(
    df_long,
    x='hour',
    y='mean_totalSum',
    color='client_NT',
    barmode='group',  # Группированные столбцы
    title='Средний чек по клиентам по часам',
    labels={
        'hour': 'Час дня',
        'mean_totalSum': 'Средний чек',
        'client_NT': 'Клиент'
    },
    text_auto='.2f'  # Автоматическое отображение значений с 2 знаками после запятой
)

# Настраиваем отображение
fig.update_layout(
    xaxis=dict(
        type='category',  # Чтобы часы отображались как категории
        title='Час дня'
    ),
    yaxis_title='Средний чек',
    legend_title='Торговая сеть',
    hovermode='x unified'  # Общая подсказка для всех столбцов при наведении
)

# Добавляем подписи значений над столбцами
fig.update_traces(
    textfont_size=12,
    textangle=0,
    textposition="outside",
    cliponaxis=False
)

fig.show()

Аналогично сформируем таблицу со средним чеком по дням недели во всех торговых сетях для сравнения:

In [42]:
mc3 = mean_check.pivot_table(
    values='totalSum',
    index='weekday',
    columns='client_NT',
    aggfunc='mean',
    observed=False
    )
mc3

client_NT      Metro c&c  Гиперглобус    Окей
weekday                                      
1.Понедельник    6744.43      3247.52 2327.74
2.Вторник        7085.23      3228.55 2341.79
3.Среда          8142.43      3228.67 2188.79
4.Четверг        9022.47      3314.41 2422.37
5.Пятница       29299.12      3557.94 2616.60
6.Суббота       13894.15      3541.05 2833.78
7.Воскресенье    5931.00      3360.12 2611.26

In [55]:
df_long = mc3.reset_index().melt(id_vars='weekday', var_name='client_NT', value_name='mean_totalSum')

# Строим barplot
fig = px.bar(
    df_long,
    x='weekday',
    y='mean_totalSum',
    color='client_NT',
    barmode='group',  # Группированные столбцы
    title='Средний чек по клиентам по часам',
    labels={
        'weekday': 'День недели',
        'mean_totalSum': 'Средний чек',
        'client_NT': 'Торговая сеть'
    },
    text_auto='.2f'  # Автоматическое отображение значений с 2 знаками после запятой
)

# Настраиваем отображение
fig.update_layout(
    xaxis=dict(
        type='category',  # Чтобы часы отображались как категории
        title='Час дня'
    ),
    yaxis_title='Средний чек',
    legend_title='Клиенты',
    hovermode='x unified'  # Общая подсказка для всех столбцов при наведении
)

# Добавляем подписи значений над столбцами
fig.update_traces(
    textfont_size=12,
    textangle=0,
    textposition="outside",
    cliponaxis=False
)

fig.show()

Сформируем сводную таблицу по среднему чеку по часам для каждого магазина торговой сети

In [63]:
mean_check.pivot_table(
    values='totalSum',
    index='hour', 
    columns=['client_NT', 'text_kktAddress'], 
    aggfunc='mean', 
    observed=False
    )

client_NT                             Metro c&c                               \
text_kktAddress г Москва, пр-кт Мира, д 211 к 1 г Москва, ул Рябиновая, д 59   
hour                                                                           
0                                           NaN                          NaN   
1                                           NaN                          NaN   
2                                           NaN                          NaN   
3                                           NaN                          NaN   
4                                           NaN                          NaN   
5                                           NaN                          NaN   
6                                           NaN                       582.90   
7                                           NaN                          NaN   
8                                      30184.29                      4936.70   
9                                      26418.88                          NaN   
10                                      9058.55                      6156.73   
11                                          NaN                      1050.39   
12                                          NaN                      6699.00   
13                                     37413.46                      7125.34   
14                                     30164.58                      4478.73   
15                                          NaN                      5967.40   
16                                          NaN                      4864.25   
17                                          NaN                      4460.68   
18                                          NaN                      5103.12   
19                                          NaN                      4087.15   
20                                          NaN                      4287.85   
21                                          NaN                      7707.20   
22                                          NaN                      1791.16   
23                                          NaN                      3892.53   

client_NT                                       \
text_kktAddress г Москва, ул Дорожная, д 1 к 1   
hour                                             
0                                          NaN   
1                                          NaN   
2                                          NaN   
3                                          NaN   
4                                          NaN   
5                                          NaN   
6                                          NaN   
7                                     54782.99   
8                                          NaN   
9                                          NaN   
10                                         NaN   
11                                         NaN   
12                                         NaN   
13                                         NaN   
14                                         NaN   
15                                         NaN   
16                                         NaN   
17                                         NaN   
18                                         NaN   
19                                         NaN   
20                                         NaN   
21                                         NaN   
22                                         NaN   
23                                         NaN   

client_NT                                                                      \
text_kktAddress г Москва, ш Ленинградское, д 71Г г Москва, км МКАД 104-й, д 6   
hour                                                                            
0                                            NaN                          NaN   
1                                            NaN                          NaN   
2                                            NaN                          NaN   
3 

Сформируем сводную таблицу по среднему чеку по дням недели для каждого магазина торговой сети

In [62]:
mean_check.pivot_table(
    values='totalSum',
    index='weekday', 
    columns=['client_NT', 'text_kktAddress'], 
    aggfunc='mean', 
    observed=False
    )

client_NT                             Metro c&c                               \
text_kktAddress г Москва, пр-кт Мира, д 211 к 1 г Москва, ул Рябиновая, д 59   
weekday                                                                        
1.Понедельник                           9058.55                      2372.96   
2.Вторник                              26418.88                      6155.26   
3.Среда                                30164.58                      4819.15   
4.Четверг                                   NaN                      5571.26   
5.Пятница                              33798.88                          NaN   
6.Суббота                                   NaN                          NaN   
7.Воскресенье                               NaN                          NaN   

client_NT                                       \
text_kktAddress г Москва, ул Дорожная, д 1 к 1   
weekday                                          
1.Понедельник                              NaN   
2.Вторник                                  NaN   
3.Среда                                    NaN   
4.Четверг                             54782.99   
5.Пятница                                  NaN   
6.Суббота                                  NaN   
7.Воскресенье                              NaN   

client_NT                                                                      \
text_kktAddress г Москва, ш Ленинградское, д 71Г г Москва, км МКАД 104-й, д 6   
weekday                                                                         
1.Понедельник                           10730.21                          NaN   
2.Вторник                                6668.59                          NaN   
3.Среда                                  9358.86                          NaN   
4.Четверг                                7434.91                          NaN   
5.Пятница                                    NaN                     28396.60   
6.Суббота                                9929.50                          NaN   
7.Воскресенье                            5931.00                          NaN   

client_NT                                                   \
text_kktAddress г Москва, Московский п, кв-л 34, д 3 стр 1   
weekday                                                      
1.Понедельник                                          NaN   
2.Вторник                                         16060.34   
3.Среда                                                NaN   
4.Четверг                                         21240.77   
5.Пятница                                              NaN   
6.Суббота                                         25788.10   
7.Воскресенье                                          NaN   

client_NT                                              \
text_kktAddress г Москва, ул Маршала Прошлякова, д 14   
weekday                                                 
1.Понедельник                                     NaN   
2.Вторник                                         NaN   
3.Среда                                           NaN   
4.Четверг                                         NaN   
5.Пятница                                    18809.88   
6.Суббота                                         NaN   
7.Воскресенье                                     NaN   

client_NT                                        \
text_kktAddress г Москва, ш Дмитровское, д 165Б   
weekday                                           
1.Понедельник                               NaN   
2.Вторник                                   NaN   
3.Среда                                     NaN   
4.Четверг                                   NaN   
5.Пятница                              31691.36   
6.Суббота                                   NaN   
7.Воскресенье                               NaN   

client_NT                                                                    Гиперглобус  \
text_kktAddress г Москва, Сосенское п, п Коммунарка, ул Александры Монаховой, д 61 стр 1   


Проведём `RFM-анализ`

Сформируем датафрейм, в котором `customer_id` не `Null`. Кроме того, возмьём лишь по одной строке `receipt_id`, так как нам необходимо узнать количество посещений

In [70]:
data_rfm = data[~data['customer_id'].isnull()] # Оставим только не нулевые `cusomer_id`
data_rfm = data_rfm.groupby('receipt_id').first() # Достаточно взять только первую строку чека
data_rfm['dateTime'] = pd.to_datetime(data_rfm['dateTime'], dayfirst=True) # Преобразуем `dateTime` в формат datetime, если вдруг он текстовый
data_rfm = data_rfm.reset_index()
snapshot_date = data_rfm['dateTime'].max() + pd.Timedelta(days=1) # Дата "снапшота" - следующий день после максимальной даты в датафрейме

In [71]:
rfm = data_rfm.groupby(['customer_id'], observed=False).agg({
    'client_NT': 'first',
    'dateTime': lambda x: (snapshot_date - pd.to_datetime(x, dayfirst=True).max()).days,
    'receipt_id': 'nunique',
    'totalSum': 'sum',  
}).rename(columns={
    'dateTime': 'Recency',      # (дней с последней покупки)
    'receipt_id': 'Frequency',  # (число уникальных чеков)
    'totalSum': 'Monetary'      # (общая сумма покупок)
});
rfm

client_NT  Recency  Frequency  Monetary
customer_id                                                                
000210750313104b1dcde90697a23cb1         Окей       53          1   1249.81
000c4642b320702ed33605722ec4bab3  Гиперглобус       25          1   4725.13
001d4e7abf6108c688d943ce99d7fc74  Гиперглобус       47          1    710.95
0023a42d6f0698d6fee2ad95121eecc3  Гиперглобус       10          2    499.64
0027021d451a8c17af41e950fe450f60         Окей       25          2   1266.14
...                                       ...      ...        ...       ...
ffdf60acbcda04db2e1d50b17432303e  Гиперглобус       57          4   6666.19
ffe6d0b5b97459e17d775c15258bd40e         Окей        4          1   1223.09
ffe6f731458ba86281ea71b474388cc8  Гиперглобус       22          1   5573.59
ffee3c7c8e93d542cb4af4b348a15fe5         Окей       54          1   1515.29
fff03395c0ba2cb78c05cf634c534ff2  Гиперглобус       15          1   1567.00

[11178 rows x 4 columns]

In [72]:
# Зададим квантили
q = [0.33, 0.66]
qr = rfm['Recency'].quantile(q).tolist()
qf = rfm['Frequency'].quantile(q).tolist()
qm = rfm['Monetary'].quantile(q).tolist()

In [73]:
# Функция для присвоения баллов (1-3)
def rfm_score(x, q, metric=False):
    if metric == 'Recency':  # Чем меньше дней, тем лучше
        if x <= q[0]: return 3
        elif x <= q[1]: return 2
        else: return 1
    else:  # Frequency и Monetary - чем больше, тем лучше
        if x <= q[0]: return 1
        elif x <= q[1]: return 2
        else: return 3

# Применяем scoring
rfm['R'] = rfm['Recency'].apply(rfm_score, q=qr, metric='Recency')
rfm['F'] = rfm['Frequency'].apply(rfm_score, q=qf)
rfm['M'] = rfm['Monetary'].apply(rfm_score, q=qm)

In [74]:
rfm

client_NT  Recency  Frequency  Monetary  \
customer_id                                                                   
000210750313104b1dcde90697a23cb1         Окей       53          1   1249.81   
000c4642b320702ed33605722ec4bab3  Гиперглобус       25          1   4725.13   
001d4e7abf6108c688d943ce99d7fc74  Гиперглобус       47          1    710.95   
0023a42d6f0698d6fee2ad95121eecc3  Гиперглобус       10          2    499.64   
0027021d451a8c17af41e950fe450f60         Окей       25          2   1266.14   
...                                       ...      ...        ...       ...   
ffdf60acbcda04db2e1d50b17432303e  Гиперглобус       57          4   6666.19   
ffe6d0b5b97459e17d775c15258bd40e         Окей        4          1   1223.09   
ffe6f731458ba86281ea71b474388cc8  Гиперглобус       22          1   5573.59   
ffee3c7c8e93d542cb4af4b348a15fe5         Окей       54          1   1515.29   
fff03395c0ba2cb78c05cf634c534ff2  Гиперглобус       15          1   1567.00   

                                  R  F  M  
customer_id                                
000210750313104b1dcde90697a23cb1  1  1  1  
000c4642b320702ed33605722ec4bab3  2  1  2  
001d4e7abf6108c688d943ce99d7fc74  1  1  1  
0023a42d6f0698d6fee2ad95121eecc3  3  2  1  
0027021d451a8c17af41e950fe450f60  2  2  1  
...                              .. .. ..  
ffdf60acbcda04db2e1d50b17432303e  1  3  3  
ffe6d0b5b97459e17d775c15258bd40e  3  1  1  
ffe6f731458ba86281ea71b474388cc8  2  1  2  
ffee3c7c8e93d542cb4af4b348a15fe5  1  1  1  
fff03395c0ba2cb78c05cf634c534ff2  3  1  1  

[11178 rows x 7 columns]

In [75]:
# Комбинированный RFM-скоринг
rfm['RFM_Score'] = rfm['R'].astype(str) + rfm['F'].astype(str) + rfm['M'].astype(str)
rfm

client_NT  Recency  Frequency  Monetary  \
customer_id                                                                   
000210750313104b1dcde90697a23cb1         Окей       53          1   1249.81   
000c4642b320702ed33605722ec4bab3  Гиперглобус       25          1   4725.13   
001d4e7abf6108c688d943ce99d7fc74  Гиперглобус       47          1    710.95   
0023a42d6f0698d6fee2ad95121eecc3  Гиперглобус       10          2    499.64   
0027021d451a8c17af41e950fe450f60         Окей       25          2   1266.14   
...                                       ...      ...        ...       ...   
ffdf60acbcda04db2e1d50b17432303e  Гиперглобус       57          4   6666.19   
ffe6d0b5b97459e17d775c15258bd40e         Окей        4          1   1223.09   
ffe6f731458ba86281ea71b474388cc8  Гиперглобус       22          1   5573.59   
ffee3c7c8e93d542cb4af4b348a15fe5         Окей       54          1   1515.29   
fff03395c0ba2cb78c05cf634c534ff2  Гиперглобус       15          1   1567.00   

                                  R  F  M RFM_Score  
customer_id                                          
000210750313104b1dcde90697a23cb1  1  1  1       111  
000c4642b320702ed33605722ec4bab3  2  1  2       212  
001d4e7abf6108c688d943ce99d7fc74  1  1  1       111  
0023a42d6f0698d6fee2ad95121eecc3  3  2  1       321  
0027021d451a8c17af41e950fe450f60  2  2  1       221  
...                              .. .. ..       ...  
ffdf60acbcda04db2e1d50b17432303e  1  3  3       133  
ffe6d0b5b97459e17d775c15258bd40e  3  1  1       311  
ffe6f731458ba86281ea71b474388cc8  2  1  2       212  
ffee3c7c8e93d542cb4af4b348a15fe5  1  1  1       111  
fff03395c0ba2cb78c05cf634c534ff2  3  1  1       311  

[11178 rows x 8 columns]

Сохраним результаты в csv-файлы

In [80]:
for trade in trades:
    rfm[rfm['client_NT'] == trade].to_csv(f'RFM {trade}.csv', index=True)

Выводы:

- Применённые функции для проведения ABC, XYZ и RFM анализов можно, при необходимости, использовать для каждого отдельного магазина, сформировав отфильтрованный датафрейм.
- ABC-анализ говорит, что товары категории:
    - A - самые важные. Это 20% товаров, которые приносят 80% выручки. Они всегда должны быть на складе. Дефицит этих товаров приводит к проседанию по выручке
    - B - требуют контроля, так как это товары-кандидаты в категорию A. Они не приносят много выручки, но лучше, чтобы они были в наличии.
    - C - товары, которые не приносят значимой выручки, их можно реже закупать и не хранить на складах. Некоторые из них можно даже исключить из ассортимента.
- XYZ-анализ говорит, что товары категории:
    - X - товары со стабильным спросом, они всегда должны быть в наличии.
    - Y - товары со средним спросом, возможно, сезонные. Закупать в большом количестве смысла нет.
    - Z - товары с редким спросом. Необходимо проанализировать поставки, если с поставками проблем не было (то есть товара было достаточно), то от таких товаров можно либо отказаться, либо поставлять по предзаказу.
- RFM-анализ говорит, что покупатели категории:
    - R - сколько дней прошло с последней покупки клиента: 1 - давно, 3 - недавно
    - F - как клиент часто покупает: 1 - редко, 3 - часто
    - M - сколько прибыли приносит клиент: 1 - мало, 3 много
    Другими словами, следует присмотреться к клиентам 133 и 233, чтобы замотивировать их чаще ходить за покупками. Например, сделать персональные предложения.
    
    Клиентов 313 и 323 также можно замотивировать ходить чаще. Например, использовать рассылку уведомлений о временных акциях на товары.
    
    Кроме того, можно обратить внимание на клиентов 222, чтобы вывести их хотя бы в одну из групп 322, 232 или 223. Подойдут и персональные предложения, и уведомления об акциях.